pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -0,0 +1,391 @@
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+type etcdrenew struct {
+	log           *logrus.Entry
+	k             adminactions.KubeActions
+	secretNames   []string
+	mode          string
+	backupSecrets map[string][]byte
+	lastRevision  int32
+}
+
+var etcdOperatorControllerConditionsExpected = map[string]operatorv1.ConditionStatus{
+	"EtcdCertSignerControllerDegraded": operatorv1.ConditionFalse,
+	"EtcdMembersAvailable":             operatorv1.ConditionTrue,
+	"NodeInstallerProgressing":         operatorv1.ConditionFalse,
+	"NodeControllerDegraded":           operatorv1.ConditionFalse,
+	"EtcdMembersProgressing":           operatorv1.ConditionFalse,
+}
+
+var etcdOperatorConditionsExpected = map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+	configv1.OperatorAvailable:   configv1.ConditionTrue,
+	configv1.OperatorProgressing: configv1.ConditionFalse,
+	configv1.OperatorDegraded:    configv1.ConditionFalse,
+}
+
+func (f *frontend) postAdminOpenShiftClusterEtcdCertificateRenew(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	err := f._postAdminOpenShiftClusterEtcdCertificateRenew(ctx, r, log)
+
+	adminReply(log, w, nil, nil, err)
+}
+
+// validate cluster is <4.9 and etcd is in expected state
+func (e *etcdrenew) validate(ctx context.Context) error {
+	s := []steps.Step{
+		steps.Action(e.validateEtcdOperatorControllersState),
+		steps.Action(e.validateEtcdOperatorState),
+		steps.Action(e.validateEtcdCertsExistsAndExpiry),
+	}
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *etcdrenew) isRenewed(ctx context.Context) error {
+	s := []steps.Step{
+		steps.Condition(e.isRevisied, 30*time.Minute, true),
+	}
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *etcdrenew) backupAndDelete(ctx context.Context) error {
+	s := []steps.Step{
+		steps.Action(e.backupEtcdSecrets),
+		steps.Action(e.deleteEtcdSecrets),
+	}
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *frontend) _postAdminOpenShiftClusterEtcdCertificateRenew(ctx context.Context, r *http.Request, log *logrus.Entry) error {
+	resourceName := chi.URLParam(r, "resourceName")
+	resourceType := chi.URLParam(r, "resourceType")
+	resourceGroupName := chi.URLParam(r, "resourceGroupName")
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.dbOpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", resourceType, resourceName, resourceGroupName)
+	case err != nil:
+		return err
+	}
```

Comment:
**Default Case is missing**, it is not mandatory to declare the default clause in a switch statement. It is entirely optional. Without it, the code will still be valid. However, as a good programming practice, you should always declare a default clause in the switch expression

Reply:
I think it's good as-is here since it's the convention we have in other adminactions, and the purpose with the switch/case here is to handle errors from cosmosdb only, not run the subsequent code.